### PR TITLE
Update SDKs for 1.2.0

### PR DESCRIPTION
**Pull request by bryce-fitzsimons**
### Features and Changes

- [x]  fix StickyBucketService [exports (2676)](https://github.com/growthbook/growthbook/issues/2676) in React SDK
- [x] refresh `stickyBucketIdentifierAttributes` when recalculating sticky buckets so that we can pick up SB docs that were previously not picked up by the SDK
- [x] make the tracking callback async, await it when navigating for url redirects
- [x] resolve some [quirks (2765)](https://github.com/growthbook/growthbook/issues/2765) with the anti flicker logic
- [x] try to prevent [navigate stampedes (2735)](https://github.com/growthbook/growthbook/issues/2735)
- [x] change default maxAge to reduce multiple exposures on phase change. expose cache settings for auto wrapper.
- [x] version bumps: sdks
- [ ] version bumps / json: GB app
- [ ] docs (pending PR review)
- [x] qa

note: version will be bumped to 1.2.0 despite the branch name

### CODE REVIEW

File: packages/sdk-js/src/GrowthBook.ts [hunk 1]
```diff
@@ -733,13 +736,22 @@ export class GrowthBook<
...
               try {
                 navigate(url);
               } catch (e) {
                 console.error(e);
               }
-            }, this._ctx.navigateDelay ?? 100);
+            });
           } else {
             try {
               navigate(url);
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:757]: wondering if we should also do something similar for non-browser contexts. I do not think we'd need this on edge fwiw.

File: packages/sdk-js/test/sticky-buckets.test.ts [hunk 2]
```diff
@@ -209,6 +209,7 @@ describe("sticky-buckets", () => {
         country: "USA",
       },
     });
+    growthbook1a.debug = true;
```
Comment (jdorn) [inline packages/sdk-js/test/sticky-buckets.test.ts:212]: Do we need this?

File: packages/sdk-js/src/auto-wrapper.ts [hunk 3]
```diff
@@ -233,20 +233,43 @@ if (
...
+        eventCallback: datalayerResolve,
+      });
+    }
 
     // Segment - analytics.js
-    window.analytics &&
-      window.analytics.track &&
-      window.analytics.track("Experiment Viewed", p);
+    if (window.analytics && window.analytics.track) {
+      window.analytics.track("Experiment Viewed", eventParams);
```
Comment (jdorn) [inline packages/sdk-js/src/auto-wrapper.ts:269]: If someone is only using Segment, previously, this trackingCallback would return void, which would cause it to use the `navigateDelay`.  Now, it will return `Promise<void>` which resolves immediately, which will cause it to immediately redirect before Segment could fire.

We should use a promise here instead that waits a hardcoded amount of time.  Segment's docs recommend using a value of `300ms` to give the tracking call enough time to finish.

File: packages/sdk-js/src/GrowthBook.ts [hunk 4]
```diff
@@ -726,20 +734,33 @@ export class GrowthBook<
...
         if (navigate) {
           if (isBrowser) {
-            this._setAntiFlicker();
-            window.setTimeout(() => {
+            this._setAntiFlicker(true);
+            // Wait for the possibly-async tracking callback, bound by min and max delays
+            Promise.all([
+              promiseTimeout(
+                this._activeTrackingCall,
+                this._ctx.maxNavigateDelay ?? 10000
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:749]: 10 seconds seems way too long here.  I was thinking it would be more like 1 second

File: packages/sdk-js/src/GrowthBook.ts [hunk 5]
```diff
@@ -726,20 +734,33 @@ export class GrowthBook<
...
+              ),
+              new Promise((resolve) =>
+                window.setTimeout(resolve, this._ctx.navigateDelay ?? 100)
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:752]: Do we want this?  If the tracking call finishes fast, we shouldn't need to wait the full 100ms

File: packages/sdk-js/test/sticky-buckets.test.ts (hunk 2 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/test/sticky-buckets.test.ts:212]: no, leftover from debugging :)

File: packages/sdk-js/src/auto-wrapper.ts (hunk 3 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/auto-wrapper.ts:269]: The SDK will always wait for at least the MAX of:
- `context.navigateDelay ?? 100`
- the tracking promise

So we could just set navigateDelay to 300

File: packages/sdk-js/src/GrowthBook.ts (hunk 4 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:749]: sure. i'm looking at it thru the lens of the folks who have been experiencing major SRM issues due to a large number of tracking calls taking over 1-2 secs

File: packages/sdk-js/src/GrowthBook.ts (hunk 5 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:752]: I put it here to protect against false positives (tracking complete after 0ms) / misconfigured tracking callbacks. It's adjustable to 0. We can remove it if you prefer though, but it does seem useful for things like the script tag

File: packages/sdk-js/src/types/growthbook.ts [hunk 6]
```diff
@@ -398,11 +399,13 @@ export type InitOptions = {
   skipCache?: boolean;
   payload?: FeatureApiResponse;
   streaming?: boolean;
+  cacheSettings?: CacheSettings;
 };
 
 export type InitSyncOptions = {
   payload: FeatureApiResponse;
   streaming?: boolean;
+  cacheSettings?: CacheSettings;
```
Comment (jdorn) [inline packages/sdk-js/src/types/growthbook.ts:408]: I don't think we need this for `initSync`, since that method doesn't use the feature repository at all.

File: packages/sdk-js/src/auto-wrapper.ts (hunk 3 above)
Comment (jdorn) [reply packages/sdk-js/src/auto-wrapper.ts:271]: I'd prefer having `navigateDelay` set to something smaller like 100, and have this Segment call specifically wait 300ms.  So they only pay that extra penalty if they are using Segment.  Otherwise, if GA finishes sooner, it doesn't have to wait.

File: packages/sdk-js/src/GrowthBook.ts [hunk 7]
```diff
@@ -214,6 +221,10 @@ export class GrowthBook<
   public initSync(options: InitSyncOptions): GrowthBook {
     this._initialized = true;
 
+    if (options.cacheSettings) {
+      configureCache(options.cacheSettings);
+    }
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:226]: Remove this from initSync

File: packages/sdk-js/src/GrowthBook.ts [hunk 8]
```diff
@@ -726,20 +743,33 @@ export class GrowthBook<
...
-            window.setTimeout(() => {
+            this._setAntiFlicker(true);
+            // Wait for the possibly-async tracking callback, bound by min and max delays
+            Promise.all([
+              promiseTimeout(
+                this._activeTrackingCall,
+                this._ctx.maxNavigateDelay ?? 3000
+              ),
+              new Promise((resolve) =>
+                window.setTimeout(resolve, this._ctx.navigateDelay ?? 300)
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:761]: Revert this to 100

File: packages/sdk-js/test/url-redirects.test.ts [hunk 9]
```diff
@@ -45,7 +45,7 @@ describe("urlRedirects", () => {
     await sleep();
 
     gb.setURL("http://www.example.com/home");
-    await sleep(100);
+    await sleep(350);
```
Comment (jdorn) [inline packages/sdk-js/test/url-redirects.test.ts:48]: Revert the changes in this file

File: packages/sdk-js/src/GrowthBook.ts [hunk 10]
```diff
@@ -683,6 +698,7 @@ export class GrowthBook<
     if (isBlocked) {
       process.env.NODE_ENV !== "production" &&
         this.log("Auto experiment blocked", { id: experiment.key });
+      this._unsetAntiFlicker();
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:701]: This doesn't feel like it belongs in this method.  Imagine you have 2 redirect experiments.  The first one is blocked, so it hits this and removes the overlay.  The next one is not blocked so it redirects, but the overlay is already gone

File: packages/sdk-js/src/GrowthBook.ts [hunk 11]
```diff
@@ -726,20 +743,33 @@ export class GrowthBook<
...
+                this._ctx.maxNavigateDelay ?? 3000
+              ),
+              new Promise((resolve) =>
+                window.setTimeout(resolve, this._ctx.navigateDelay ?? 300)
+              ),
+            ]).then(() => {
               try {
-                navigate(url);
+                if (!this._isNavigating) {
+                  this._isNavigating = true;
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:766]: This needs to get unset after the navigation finishes.  Especially important for SPAs where the navigate callback doesn't do a full page load.

File: packages/sdk-js/src/GrowthBook.ts [hunk 12]
```diff
@@ -1483,7 +1513,8 @@ export class GrowthBook<
     }
 
     // 14. Fire the tracking callback
-    this._track(experiment, result);
+    // Store the promise in case we're awaiting it (ex: browser url redirects)
+    this._activeTrackingCall = this._track(experiment, result);
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1517]: I don't like the idea of storing this globally.  Seems like it might open us up to race conditions.  Instead, we could include this promise in the return value of the function.

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:1513]: pinged you, but I kind of disagree, unless I'm misunderstanding your suggestion:
- We'd need to add something like `result.trackingCall` which may not serialize well for people consuming/logging `result` directly. We might break existing integrations.
- the global `_activeTrackingCall` is set and consumed in the same exe cycle since `_runAutoExperiment()` is synchronous, so its hard to imagine a race condition occurring.

File: packages/sdk-js/src/GrowthBook.ts (hunk 11 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:762]: Do you mean something like
```
navigate(url).then(() => this._isNavigating = false);
```
One possible problem is that navigate() may not be async or at least not trackable (such as `window.location.replace(url)`), so it would get unset instantly. Maybe that's ok though, but then it doesn't do much to dedupe.

Or perhaps we make the default navigate function return a promise with an arbitrary 1sec timeout?

File: packages/sdk-js/src/GrowthBook.ts (hunk 10 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:697]: Sure we can remove. The flip side though is the anti-flicker stays up far too long (3000ms by default) when only a single experiment triggers but is blocked.

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:697]: there are a few other bail-out scenarios where i'm calling `_unsetAntiFlicker();`:
- redirect variant does not actually redirect (ex: control)
- invalid urlPatterns (rare)
- already on the target page (rare)

the same double experiment scenario you mentioned earlier could happen here too, right?

File: packages/sdk-js/src/GrowthBook.ts (hunk 12 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts]: discussed. we decided to return the tracking call locally in the private _run() method without affecting public run()

File: packages/sdk-js/src/GrowthBook.ts (hunk 11 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:768]: move anti flicker to _updateAllAutoExperiments, outside of loop

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:768]: no anti-flicker in constructor, just as part of the (default) navigate function.

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:768]: no anti-flicker for dom changes. only for default navigate function. make sure hooks work...

File: packages/sdk-js/src/auto-wrapper.ts [hunk 13]
```diff
@@ -266,6 +297,44 @@ gb.init({
     dataContext.noStreaming ||
     windowContext.backgroundSync === false
   ),
+  cacheSettings: windowContext.cacheSettings,
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/auto-wrapper.ts:300]: if antiFlicker set, hide the screen before init(). Otherwise we're waiting for features to load

**Comment by bryce-fitzsimons** [comment]
todo: add unit test for tracking call being async

**Comment by bryce-fitzsimons** [comment]
unit test: navigate stampede. custom navigate function, make sure we aren't running exp multiple times / re-navigating

File: packages/sdk-js/src/GrowthBook.ts [hunk 14]
```diff
@@ -1741,35 +1787,27 @@ export class GrowthBook<
...
+        delay: 100,
+      };
     } else if (isBrowser) {
-      return (url: string) => {
-        window.location.replace(url);
+      return {
+        navigate: (url: string) => {
+          window.location.replace(url);
+        },
+        delay: 0,
```
Comment (jdorn) [inline packages/sdk-js/src/GrowthBook.ts:1804]: These delays are backwards?

File: packages/sdk-js/src/auto-wrapper.ts [hunk 15]
```diff
@@ -205,6 +209,46 @@ function getAttributes() {
...
 
+function setAntiFlicker() {
+  window.clearTimeout(antiFlickerTimeout);
+
+  const timeoutMs =
+    windowContext.antiFlickerTimeout ??
+    (dataContext.antiFlickerTimeout
+      ? parseInt(dataContext.antiFlickerTimeout)
+      : null) ??
+    3500;
```
Comment (jdorn) [inline packages/sdk-js/src/auto-wrapper.ts:220]: If `dataContext.antiFlickerTimeout` is not parseable as an int (e.g. `"a"`), this will result in `NaN`.  This means, the delay will be treated as `0ms` instead of `3500ms`.

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/auto-wrapper.ts:220]: that seems like a user error, but can add some guards

**Review by jdorn — APPROVED** [review]
